src/crimson/osd/osd.h

```diff
@@ -137,6 +146,61 @@ class OSD : public ceph::net::Dispatcher,
   void write_superblock(ceph::os::Transaction& t);
   seastar::future<> read_superblock();
 
+  bool require_mon_peer(ceph::net::Connection *conn, Ref<Message> m);
+
+  seastar::future<Ref<PG>> handle_pg_create_info(
+    std::unique_ptr<PGCreateInfo> info);
+
+  template <typename C, typename F, typename G>
+  seastar::future<> handle_batch_pg_message_with_missing_handler(
+    const C &c,
+    F &&f,
+    G &&on_missing_pg) {
+    using mapped_type = decltype(*(c.begin()));
```

Comment:
It appears this template deals with regular containers as:

* `MOSDPGQuery::pg_list` is `std::map<spg_t, pg_query_t>`,
* `MOSDPGCreate2::pgs` – `map<spg_t,pair<epoch_t,utime_t>>` while
* `MOSDPGNotify::get_pg_list()` returns `std::vector<std::pair<pg_notify_t,PastIntervals>>` .

Maybe instead of the `decltype` we could use something like:

```cpp
      using mapped_type = typename C::value_type;
```